plugins/nordic-hid/fu-nordic-hid-cfg-channel.c

```diff
@@ -0,0 +1,1054 @@
+
+	if (!fu_nordic_hid_cfg_channel_cmd_send(self,
+						self->peer_id,
+						"dfu",
+						"fwinfo",
+						CONFIG_STATUS_FETCH,
+						NULL,
+						0,
+						error))
+		return FALSE;
+	if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
+		return FALSE;
+
+	/* parsing fwinfo answer */
+	/* TODO: add banks amount into quirk */
+	if (res->data[0] > 1) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "invalid flash area returned by device");
+		return FALSE;
+	}
+	/* set the target flash ID area */
+	self->flash_area_id = res->data[0] ^ 1;
+	if (!fu_common_read_uint32_safe(res->data,
+					REPORT_SIZE,
+					0x01,
+					&self->flashed_image_len,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe(res->data,
+					REPORT_SIZE,
+					0x07,
+					&ver_rev,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	if (!fu_common_read_uint32_safe(res->data,
+					REPORT_SIZE,
+					0x09,
+					&ver_build_nr,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+	version = g_strdup_printf("%u.%u.%u.%u", res->data[4], res->data[5], ver_rev, ver_build_nr);
+	fu_device_set_version(FU_DEVICE(self), version);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_dfu_reboot(FuNordicDeviceCfgChannel *self, GError **error)
+{
+	g_autoptr(FuNordicCfgChannelMsg) res = g_new0(FuNordicCfgChannelMsg, 1);
+
+	if (!fu_nordic_hid_cfg_channel_cmd_send(self,
+						self->peer_id,
+						"dfu",
+						"reboot",
+						CONFIG_STATUS_FETCH,
+						NULL,
+						0,
+						error))
+		return FALSE;
+	if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
+		return FALSE;
+	if (res->data_len != 1 || res->data[0] != 0x01) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "reboot data was invalid");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_nordic_hid_cfg_channel_dfu_sync(FuNordicDeviceCfgChannel *self,
+				   FuNordicCfgChannelDfuInfo *dfu_info,
+				   guint8 expecting_state,
+				   GError **error)
+{
+	guint tries = FU_NORDIC_HID_CFG_CHANNEL_RETRIES;
+	g_autoptr(FuNordicCfgChannelMsg) res = g_new0(FuNordicCfgChannelMsg, 1);
+
+	while (TRUE) {
+		if (!fu_nordic_hid_cfg_channel_cmd_send(self,
+							self->peer_id,
+							"dfu",
+							"sync",
+							CONFIG_STATUS_FETCH,
+							NULL,
+							0,
+							error))
+			return FALSE;
+		if (!fu_nordic_hid_cfg_channel_cmd_receive(self, CONFIG_STATUS_SUCCESS, res, error))
+			return FALSE;
+		if (res->data_len != 0x0F) {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_NOT_SUPPORTED,
+					    "incorrect length of sync reply");
+			return FALSE;
+		}
+		if (res->data[0] == expecting_state)
+			break;
+
+		if (tries-- == 0) {
+			g_set_error_literal(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_NOT_SUPPORTED,
+					    "too lot of sync attempts");
```

Comment:
`too many sync attempts` perhaps -- and I do wonder if fu_device_retry_full might be better to use here.